streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractProcessorContext.java

```diff
@@ -45,7 +45,6 @@
     private boolean initialized;
     protected ProcessorRecordContext recordContext;
     protected ProcessorNode<?, ?, ?, ?> currentNode;
-    private long currentSystemTimeMs;
```

Comment:
@guozhangwang -- I am just wondering if removing this variable (what I think conceptually make sense) might have an perf impact as it seems `currentSystemTimeMs` acts as some kind of cache -- and instead of using the cached time, we would call `time.currentTimeMillis()` in `ProcessorContextImpl#forwardInternal()` now.

Thoughts?

Reply:
I think its rationale is actually to "not advance" the system time while in the middle of processing a record, i.e. the `system time` would just be the time when we start processing / punctuating, and it would not change throughout the process / punctuate.

Reply:
@rohitrmd did you see Guozhang comment? Seems we should not remove this variable (what is a little unfortunate, but I guess desirable behavior) -- maybe we can rename it to `cachedSystemTimeMs`?

I guess we want the caching for the `GlobalProcessorContext`, too? (For this case, we would need to update the JavaDocs of the `ProcessorContext` interface accordingly.)

We should also add a test for the corresponding contest-test classes that they don't advance system time automatically, but only in the advance is triggered explicitly.

Reply:
@mjsax do you recommend adding method to AbstractProcessorContext to set cachedSystemTimeMs (setCachedSystemTimeMs)?

Reply:
@mjsax also I am confused about what time do we want to return when currentSystemTimeMs() is called from the ProcessorContext hierarchy. If it's called from AbstractProcessorContext object reference, should we return cachedSystemTimeMs, and if it's called from ProcessorContextImpl object reference, should we return streamTask.currentSystemTimeMs()? Or we want completely new method to return the cached time when ProcessorContextImpl#forwardInternal() is called?

Reply:
If there is an advantage to add the method to the `InternalProcessorContext ` interface, sure why not. Don't think it's a big difference, and because it all internal code, we can change it anytime in the future.

I think we should always return the cached time.

Reply:
@mjsax can you please explain again what is expected now as for me this is contrary to initial KIP. What I have understood from KIP, we wanted to return system time from Stream Task. Considering changes in this pr, 

1. We added one time field in InternalMockProcessorContext which we return when currentSystemTime() is called. 
2. When ProcessorContextImpl's currentSystemTime() method is called, we return time from streamTask's time field. 
3. We also added time field in GlobalProcessorContextImpl which we return from currentSystemTime().
4. If we add new cachedSystemTimeMs field in AbstractProcessorContext, when do you want to return this field? Are earlier changes not valid to return time from StreamTask?

Reply:
The main motivation to add `ProcessorContext#currentSystemTime()` was to be able to return the mocked wall-clock time in `TopologyTestDriver`. Even if we return the cached time from `AbstractProcessorContext`, we will be able to return the mocked time, as we update the cached time based on the mocked time in `TopologyTestDriver`.

- `InternalMockProcessorContext` is just for our own unit testing -- it's fine to add the new `Time` field, it's not a public facing change anyway
- Originally we changed `ProcessorContextImpl` because we remove `AbstractProcessorContext#currentSystemTime()` -- as suggested by Guozhang, we should keep the cached time in `AbstractProcessorContext()`, and thus we don't need `ProcessorContextImpl#currentSystemTime()` any longer.
- `GlobalProcessorContextImpl` is a different code path, and thus the changes of this PR are fine

> If we add new cachedSystemTimeMs field in AbstractProcessorContext, when do you want to return this field?

Yes, we want to return this field. (We get this behaviro by adding back `AbstractProcessorContext#currentSystemTime()` (and the cached time in this class) and removing `ProcessorContextImpl#currentSystemTime()`.

> Are earlier changes not valid to return time from StreamTask?

Yes, I think we can revert all changes from `StreamTask`.

Does this make sense?

Reply:
@mjsax yes, thank you for the explanation. made changes as per review.